src/main/java/net/imglib2/img/basictypeaccess/Accesses.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * #%L
+ * ImgLib2: a general-purpose, multidimensional image processing library.
+ * %%
+ * Copyright (C) 2009 - 2018 Tobias Pietzsch, Stephan Preibisch, Stephan Saalfeld,
+ * John Bogovic, Albert Cardona, Barry DeZonia, Christian Dietz, Jan Funke,
+ * Aivar Grislis, Jonathan Hale, Grant Harris, Stefan Helfrich, Mark Hiner,
+ * Martin Horn, Steffen Jaensch, Lee Kamentsky, Larry Lindsey, Melissa Linkert,
+ * Mark Longair, Brian Northan, Nick Perry, Curtis Rueden, Johannes Schindelin,
+ * Jean-Yves Tinevez and Michael Zinsmaier.
+ * %%
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package net.imglib2.img.basictypeaccess;
+
+/**
+ * Utility and helper methods for accesses ({@link ByteAccess} etc).
+ *
+ * @author Philipp Hanslovsky
+ */
+public interface Accesses
+{
+
+	public static void copy( final ByteAccess source, final ByteAccess target, final int size )
+	{
+		copy( source, target, 0, size );
+	}
+
+	public static void copy( final ByteAccess source, final ByteAccess target, final int start, final int stop )
+	{
+		for ( int index = start; index < stop; ++index )
+		{
+			target.setValue( index, source.getValue( index ) );
```

Comment:
In cases where we have an `ArrayDataAccess`, we can do better performance-wise, by calling `getCurrentStorageArray()` followed by `System.arraycopy`. I suggest adding this as a special case somehow.

Reply:
I wonder if this is pre-mature optimization. If copying from one access to another access is a bottleneck, there is probably something wrong with the code. Also, `ArrayDataAccess.getCurrentStorageArray()` returns `Object` and is not even guaranteed to return array.
 A way around this would be to add
```java
public default void copyInto( ByteAccess target )
{
// do the copy as in the proposed methods
}
```
to the `ByteAccess` etc. interfaces and override with special case implementations for `ByteArray` etc.

Reply:
FWIW, I made a little "benchmark":
```java
import net.imglib2.img.basictypeaccess.ByteAccess;
import net.imglib2.img.basictypeaccess.array.ByteArray;

public class ByteAccessCopyTest
{

	public static void main( final String args[] )
	{
		final int iterations = 100000;

		final int size = 1000000;

		final ByteArray source = new ByteArray( size );
		final ByteArray target = new ByteArray( size );

		for ( int i = 0; i < iterations; ++i )
		{
			copy( source, target, size );
		}

		final long tStartAccessCopy = System.nanoTime();
		for ( int i = 0; i < iterations; ++i )
		{
			copy( source, target, size );
		}
		final long tStopAccessCopy = System.nanoTime();
		final long tAccessCopy = tStopAccessCopy - tStartAccessCopy;

		for ( int i = 0; i < iterations; ++i )
		{
			System.arraycopy( source.getCurrentStorageArray(), 0, target.getCurrentStorageArray(), 0, iterations );
		}

		final long tStartArraycopy = System.nanoTime();
		for ( int i = 0; i < iterations; ++i )
		{
			System.arraycopy( source.getCurrentStorageArray(), 0, target.getCurrentStorageArray(), 0, iterations );
		}
		final long tStopArraycopy = System.nanoTime();
		final long tArraycopy = tStopArraycopy - tStartArraycopy;

		System.out.println( "access copy: " + tAccessCopy * 1.0 / iterations );
		System.out.println( "array  copy: " + tArraycopy * 1.0 / iterations );

	}

	private static final void copy( final ByteAccess source, final ByteAccess target, final int size )
	{
		for ( int i = 0; i < size; ++i )
		{
			target.setValue( i, source.getValue( i ) );
		}
	}

}
```

```
access copy: 111722.59448
array  copy: 4472.66502
```
Might be worth having those special case implementations. Is there a good way to check if an `Object` is instance of primtiive type other than doing `obj instance of byte[] || obj instance of char[]` etc.?

Reply:
I expanded your performance test to include `ShortArray` and `FloatArray` as well mixed into one scope, and speed remains equivalent to `System.arraycopy`. Pretty impressive JIT!